### PR TITLE
Updated AAD auth to use our app and tenant based auth

### PR DESCRIPTION
**Pull request by wbreza**
Updated AAD auth to use our app and tenant based auth

### CODE REVIEW

File: ProvisionSample/Program.cs [hunk 1]
```diff
@@ -588,29 +594,65 @@ static void Main(string[] args)
             return client;
         }
 
+        static async Task<IEnumerable<string>> GetTenantIdsAsync(string commonToken)
+        {
+            using (var httpClient = new HttpClient())
+            {
+                httpClient.DefaultRequestHeaders.Add("Authorization", "Bearer " + commonToken);
+                var response = await httpClient.GetStringAsync("https://management.azure.com/tenants?api-version=2016-01-29");
```
Comment (druttka) [inline ProvisionSample/Program.cs:602]: I see that GetAzureAccessTokenAsync checks for this returning an empty set, and in that case throws an exception that it couldn't get the tenant. 

Let's consider the exceptions that could occur in GetTenantIdsAsync (network failure, non-success response codes, success responses that don't have the expected "value"/"tenantId" tokens) - do we want to catch those and put them down the same flow of "couldn't get the tenant," or have them throw up the stack as is?

Comment (wbreza) [inline ProvisionSample/Program.cs:602]: Definitely agree for production use cases.  This is only a sample app to give basic guidance to our customers and demonstrates how to get the AAD tokens / tenants.

File: ProvisionSample/Program.cs [hunk 2]
```diff
@@ -588,29 +594,65 @@ static void Main(string[] args)
...
+        static async Task<string> GetAzureAccessTokenAsync()
         {
-            // Follow instructions here to setup your tenants provisioning app: https://azure.microsoft.com/en-us/documentation/articles/resource-group-create-service-principal-portal/#get-access-token-in-code
+            if (!string.IsNullOrWhiteSpace(azureToken))
+            {
+                return azureToken;
+            }
+
+            var commonToken = GetCommonAzureAccessToken();
+            var tenantId = (await GetTenantIdsAsync(commonToken.AccessToken)).FirstOrDefault();
```
Comment (mattmazzola) [inline ProvisionSample/Program.cs:622]: Is `FirstOrDefault()` used based on assumption that most people using the app only have a single tenant or is there always only 1 tenant by design of Azure?  I thought tenant was equivalent of directories and there could be multiple meaning we might select the tenant id of the one the consumer does not intend to use.

**Comment by mattmazzola** [comment]
Looks good to me.

One thing to consider is how you store state of tokens.  For instance, the class member `azureToken`.  From what I've seen, best practice for ADAL usage is to not store any token data manually outside of ADAL, but to always get it through their `aquiteToken*` methods which access your implementation of TokenCache.  This ensures consistent access and likely more secure access (depending on who you write TokenCache) and prevents mistakes from ADAL being out of sync with local memory state.

That being said, it's just sample so might not be worth the refactor.

I remember you and Andy talking about this yesterday, but I didn't catch the details.  What is the reasoning for having first get the token through the common authority and then get another token which is specific to the tenant?

File: ProvisionSample/Program.cs (hunk 2 above)
Comment (wbreza) [inline ProvisionSample/Program.cs:622]: Agreed for production use case - good enough for a sample.

**Comment by wbreza** [comment]
@mattmazzola We need a tenant based token to call into ARM.  When you authenticate with the `common` tenant you don't get the users tenant information.  We need a token to get the users tenants and from there can call into ADAL to request the tenant based token.

**Pull request merged**